cpp/src/arrow/compute/exec/options.h

```diff
@@ -397,23 +397,33 @@ class ARROW_EXPORT HashJoinNodeOptions : public ExecNodeOptions {
 /// This node will output one row for each row in the left table.
 class ARROW_EXPORT AsofJoinNodeOptions : public ExecNodeOptions {
  public:
-  AsofJoinNodeOptions(FieldRef on_key, FieldRef by_key, int64_t tolerance)
-      : on_key(std::move(on_key)), by_key(std::move(by_key)), tolerance(tolerance) {}
+  AsofJoinNodeOptions(FieldRef on_key, const FieldRef& by_key, int64_t tolerance)
+      : on_key(std::move(on_key)), by_key(), tolerance(tolerance) {
+    this->by_key.push_back(std::move(by_key));
+  }
+
+  AsofJoinNodeOptions(FieldRef on_key, std::vector<FieldRef> by_key, int64_t tolerance)
+      : on_key(std::move(on_key)), by_key(by_key), tolerance(tolerance) {}
+
+  // resolves ambiguity between previous constructors when initializer list is given
+  AsofJoinNodeOptions(FieldRef on_key, std::initializer_list<FieldRef> by_key,
+                      int64_t tolerance)
+      : on_key(std::move(on_key)), by_key(by_key), tolerance(tolerance) {}
 
   /// \brief "on" key for the join. Each
   ///
-  /// All inputs tables must be sorted by the "on" key. Inexact
-  /// match is used on the "on" key. i.e., a row is considiered match iff
+  /// All inputs tables must be sorted by the "on" key. Must be a single field of a common
+  /// type. Inexact match is used on the "on" key. i.e., a row is considered match iff
   /// left_on - tolerance <= right_on <= left_on.
-  /// Currently, "on" key must be an int64 field
+  /// Currently, the "on" key must be of an integer or timestamp type
```

Comment:
by integer, do you mean int64, int32 or both?

Reply:
I mean all of `{u,}int{8,16,32,64}_t`. See the [supported types](https://github.com/apache/arrow/pull/13880/files#diff-5789a42aebc3bd0f5a6db687c78ab0c3eef3e316982f99fe9e0ea21fabed354cR1000-R1040) for a complete list.